core/templates/dev/head/pages/exploration_player/ImagePreloaderServiceSpec.js

```diff
@@ -0,0 +1,356 @@
+        exploration_id: '1',
+        filename: 'sIOFeedback.png'
+      });
+    requestUrl4 = UrlInterpolationService.interpolateUrl(
+      '/imagehandler/<exploration_id>/<filename>', {
+        exploration_id: '1',
+        filename: 's6Hint1.png'
+      });
+  }));
+
+  it('should maintain the correct number of download requests in queue',
+    function() {
+      $httpBackend.expect('GET', requestUrl1).respond(201, 'image data 1');
+      $httpBackend.expect('GET', requestUrl2).respond(201, 'image data 2');
+      $httpBackend.expect('GET', requestUrl3).respond(201, 'image data 3');
+      $httpBackend.expect('GET', requestUrl4).respond(201, 'image data 4');
+      var exploration = eof.createFromBackendDict(explorationDict);
+      ips.init(exploration);
+      ips.kickOffImagePreloader(exploration.getInitialState().name);
+      expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(3);
+      expect(ips.isLoadingImageFile('sIMChoice1.png')).toBe(true);
+      expect(ips.isLoadingImageFile('sIMChoice2.png')).toBe(true);
+      expect(ips.isLoadingImageFile('sIOFeedback.png')).toBe(true);
+      expect(ips.isLoadingImageFile('s6Hint1.png')).toBe(false);
+      $httpBackend.flush(1);
+      expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(3);
+      $httpBackend.flush(1);
+      expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(2);
+      $httpBackend.flush(1);
+      expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(1);
+      $httpBackend.flush(1);
+      expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(0);
+      expect(ips.isLoadingImageFile('sIMChoice1.png')).toBe(false);
+      expect(ips.isLoadingImageFile('sIMChoice2.png')).toBe(false);
+      expect(ips.isLoadingImageFile('sIOFeedback.png')).toBe(false);
+      expect(ips.isLoadingImageFile('s6Hint1.png')).toBe(false);
+    });
+
+  it('should properly restart pre-loading from a new state', function() {
+    var exploration = eof.createFromBackendDict(explorationDict);
+    ips.init(exploration);
+    ips.kickOffImagePreloader(exploration.getInitialState().name);
+    expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(3);
+    ips.restartImagePreloader('State 6');
+    expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(1);
+    expect(ips.isLoadingImageFile('s6Hint1.png')).toBe(true);
+  });
+
+  it('should verify that preloader starts when state changes', function() {
+    var exploration = eof.createFromBackendDict(explorationDict);
+    ips.init(exploration);
+    ips.kickOffImagePreloader(exploration.getInitialState().name);
+    expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(3);
+    expect(ips.isLoadingImageFile('s6Hint1.png')).toBe(false);
+    ips.onStateChange('State 6');
+    expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(1);
+    expect(ips.isLoadingImageFile('s6Hint1.png')).toBe(true);
+  });
```

Comment:
What is the difference between this tests and the test before? T6hey differ only in one line.

Reply:
Oh ok, I see now.